src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -34,8 +34,11 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
         private readonly DocumentVersionCache _documentVersionCache;
         private readonly ILogger _logger;
 
-        // Caches the last response per-document to potentially save on computation costs.
-        private readonly MemoryCache<DocumentUri, SemanticTokensCacheResponse> _cachedResponses = new();
+        // Cache containing semantic token range responses on a per-document + semantic version basis.
+        // Stores URIs mapping to semantic versions, which in turn map to ranges and tokens.
+        private const int MaxSemanticVersionsPerDoc = 5;
+        private const int MaxRangesPerSemanticVersion = 1000;
+        private readonly MemoryCache<DocumentUri, MemoryCache<VersionStamp, MemoryCache<Range, SemanticTokens>>> _cachedResponses = new();
```

Comment:
IF we extracted the memory cache of memory caches into its own type would that help in abstracting away some of the complexity of having a dictionary of dictionaries. i.e. `SemanticTokensCache _cachedResponses = new()`?